File: ddd/tests/fixtures/operation-error-set/typescript-companion-workspace/billing-domain/src/shadowed.ts
/**
 * An application type named `Result`. This module never imports the configured
 * language-support result, so the spelling alone is all that could establish the
 * standard result identity here.
 */

export type Result<T, E> =
  | { readonly ok: true; readonly value: T }
  | { readonly ok: false; readonly error: E };

export type IssueInvoiceError = "already-issued" | "empty-lines";
export type OpenInvoiceError = "negative-amount" | "missing-customer";

const invoiceBrand: unique symbol = Symbol("Invoice");

export type Invoice = {
  readonly [invoiceBrand]: true;
  issue(): Result<void, IssueInvoiceError>;
};

export const Invoice = {
  open(): Result<Invoice, OpenInvoiceError> {
    const instance: Invoice = {
      [invoiceBrand]: true,
      issue() {
        return { ok: true, value: undefined };
      },
    };
    return { ok: true, value: instance };
  },
};

File: ddd/tests/fixtures/operation-error-set/rust-workspace/billing-domain/src/shadowed.rs
//! An application type named `Result`. Inside this module it shadows the
//! prelude, so the spelling alone never establishes the standard result.

pub struct Result<T, E>(pub T, pub E);

pub enum IssueInvoiceError {
    AlreadyIssued,
    EmptyLines,
}

pub enum OpenInvoiceError {
    NegativeAmount,
    MissingCustomer,
}

pub struct Invoice;

impl Invoice {
    pub fn issue(&mut self) -> Result<(), IssueInvoiceError> {
        Result((), IssueInvoiceError::AlreadyIssued)
    }

    pub fn open() -> Result<Self, OpenInvoiceError> {
        Result(Self, OpenInvoiceError::NegativeAmount)
    }
}
